File: sdks/csharp/src/Verdocs.Sdk/Lists.cs
using Verdocs.Models;

namespace Verdocs;

/// <summary>
/// Constant lists shared with the other Verdocs SDKs: the known field types, their default
/// builder dimensions, and the full permission set.
/// </summary>
public static class Lists
{
    /// <summary>
    /// Every known field type, in the js-sdk's order. "textarea" is deprecated (use "textbox"
    /// with multiline enabled) but still appears in stored templates.
    /// </summary>
    public static IReadOnlyList<string> FieldTypes { get; } =
    [
        FieldType.Textbox,
        FieldType.Signature,
        FieldType.Initial,
        FieldType.Date,
        FieldType.Dropdown,
        FieldType.Timestamp,
        FieldType.Textarea,
        FieldType.Checkbox,
        FieldType.Radio,
        FieldType.Attachment,
        FieldType.Payment,
    ];

    /// <summary>Default field widths (in document units) used when placing new fields.</summary>
    public static IReadOnlyDictionary<string, double> DefaultFieldWidths { get; } = new Dictionary<string, double>
    {
        [FieldType.Signature] = 71,
        [FieldType.Initial] = 71,
        [FieldType.Date] = 75,
        [FieldType.Timestamp] = 130,
        [FieldType.Textbox] = 150,
        [FieldType.Textarea] = 150,
        [FieldType.Checkbox] = 14,
        [FieldType.Radio] = 14,
        [FieldType.Dropdown] = 85,
        [FieldType.Attachment] = 24,
        [FieldType.Payment] = 24,
    };

    /// <summary>Default field heights (in document units) used when placing new fields.</summary>
    public static IReadOnlyDictionary<string, double> DefaultFieldHeights { get; } = new Dictionary<string, double>
    {
        [FieldType.Signature] = 36,
        [FieldType.Initial] = 36,
        [FieldType.Date] = 15,
        [FieldType.Timestamp] = 15,
        [FieldType.Textbox] = 15,
        [FieldType.Textarea] = 41,
        [FieldType.Checkbox] = 14,
        [FieldType.Radio] = 14,
        [FieldType.Dropdown] = 20,
        [FieldType.Attachment] = 24,
        [FieldType.Payment] = 24,
    };

    /// <summary>Every permission that can be granted to a profile or group.</summary>
    public static IReadOnlyList<string> AllPermissions { get; } =
    [
        "template:creator:create:public",
        "template:creator:create:org",
        "template:creator:create:personal",
        "template:creator:delete",
        "template:creator:visibility",
        "template:member:read",
        "template:member:write",
        "template:member:delete",
        "template:member:visibility",
        "owner:add",
        "owner:remove",
        "admin:add",
        "admin:remove",
        "member:view",
        "member:add",
        "member:remove",
        "org:create",
        "org:view",
        "org:update",
        "org:delete",
        "org:transfer",
        "org:list",
        "envelope:create",
        "envelope:cancel",
        "envelope:view",
    ];
}

File: sdks/python/src/verdocs/models/lists.py
"""Field-placement constants and permission names (js-sdk: Lists.ts)."""

from __future__ import annotations

from .base import FieldType

FIELD_TYPES: tuple[FieldType, ...] = (
    "textbox",
    "signature",
    "initial",
    "date",
    "dropdown",
    "timestamp",
    # Deprecated upstream: use textbox with multiline instead. Kept so the
    # list still covers everything the server accepts.
    "textarea",
    "checkbox",
    "radio",
    "attachment",
    "payment",
)

# Default sizes (in document units) applied when a field is first placed.
DEFAULT_FIELD_WIDTHS: dict[FieldType, int] = {
    "signature": 71,
    "initial": 71,
    "date": 75,
    "timestamp": 130,
    "textbox": 150,
    "textarea": 150,
    "checkbox": 14,
    "radio": 14,
    "dropdown": 85,
    "attachment": 24,
    "payment": 24,
}

DEFAULT_FIELD_HEIGHTS: dict[FieldType, int] = {
    "signature": 36,
    "initial": 36,
    "date": 15,
    "timestamp": 15,
    "textbox": 15,
    "textarea": 41,
    "checkbox": 14,
    "radio": 14,
    "dropdown": 20,
    "attachment": 24,
    "payment": 24,
}

ALL_PERMISSIONS: tuple[str, ...] = (
    "template:creator:create:public",
    "template:creator:create:org",
    "template:creator:create:personal",
    "template:creator:delete",
    "template:creator:visibility",
    "template:member:read",
    "template:member:write",
    "template:member:delete",
    "template:member:visibility",
    "owner:add",
    "owner:remove",
    "admin:add",
    "admin:remove",
    "member:view",
    "member:add",
    "member:remove",
    "org:create",
    "org:view",
    "org:update",
    "org:delete",
    "org:transfer",
    "org:list",
    "envelope:create",
    "envelope:cancel",
    "envelope:view",
)
